ouroboros-consensus-cardano/src/tools/Cardano/Tools/ImmDBServer/MiniProtocols.hs

```diff
@@ -0,0 +1,231 @@
+            [ mkMiniProtocol
+                N2N.keepAliveMiniProtocolNum
+                N2N.keepAliveProtocolLimits
+                keepAliveProt
+            , mkMiniProtocol
+                N2N.chainSyncMiniProtocolNum
+                N2N.chainSyncProtocolLimits
+                chainSyncProt
+            , mkMiniProtocol
+                N2N.blockFetchMiniProtocolNum
+                N2N.blockFetchProtocolLimits
+                blockFetchProt
+            , mkMiniProtocol
+                N2N.txSubmissionMiniProtocolNum
+                N2N.txSubmissionProtocolLimits
+                txSubmissionProt
+            ]
+          where
+            Consensus.N2N.Codecs {
+                cKeepAliveCodec
+              , cChainSyncCodecSerialised
+              , cBlockFetchCodecSerialised
+              } =
+              Consensus.N2N.defaultCodecs codecCfg blockVersion encAddr decAddr version
+
+            keepAliveProt  =
+                MuxPeer nullTracer cKeepAliveCodec
+              $ keepAliveServerPeer keepAliveServer
+            chainSyncProt  =
+                MuxPeerRaw \channel ->
+                withRegistry
+              $ runPeer nullTracer cChainSyncCodecSerialised channel
+              . chainSyncServerPeer
+              . chainSyncServer immDB ChainDB.getSerialisedHeaderWithPoint
+            blockFetchProt =
+                MuxPeerRaw \channel ->
+                withRegistry
+              $ runPeer nullTracer cBlockFetchCodecSerialised channel
+              . blockFetchServerPeer
+              . blockFetchServer immDB ChainDB.getSerialisedBlockWithPoint
+            txSubmissionProt =
+                -- never reply, there is no timeout
+                MuxPeerRaw \_channel -> forever $ threadDelay 10
+
+        mkMiniProtocol miniProtocolNum limits proto = MiniProtocol {
+            miniProtocolNum
+          , miniProtocolLimits = limits N2N.defaultMiniProtocolParameters
+          , miniProtocolRun    = ResponderProtocolOnly proto
+          }
+
+chainSyncServer ::
+     forall m blk a. (IOLike m, HasHeader blk)
+  => ImmutableDB m blk
+  -> BlockComponent blk (ChainDB.WithPoint blk a)
+  -> ResourceRegistry m
+  -> ChainSyncServer a (Point blk) (Tip blk) m ()
+chainSyncServer immDB blockComponent registry = ChainSyncServer $ do
+    follower <- newImmutableDBFollower
+    runChainSyncServer $
+      chainSyncServerForFollower nullTracer getImmutableTip follower
+  where
+    newImmutableDBFollower :: m (Follower m blk (ChainDB.WithPoint blk a))
+    newImmutableDBFollower = do
+        varIterator <-
+          newTVarIO =<< ImmutableDB.streamAll immDB registry blockComponent
+
+        let followerInstructionBlocking = do
+              iterator <- readTVarIO varIterator
+              ImmutableDB.iteratorNext iterator >>= \case
+                ImmutableDB.IteratorExhausted -> do
+                  ImmutableDB.iteratorClose iterator
+                  throwIO ReachedImmutableTip
```

Comment:
No, only the client can terminate the ChainSync protocol by sending `MsgDone`. Instead of throwing an exception, we could wait forever, but that seems a bit more implicit/potentially confusing ("why is nothing making progress anymore?").